File: CommandPattern/java_/RemoteControl/RemoteControl.java
package CommandPattern.java_.RemoteControl;

import java.util.*;
import CommandPattern.java_.Command.Command;

public class RemoteControl {
    private Command[] onCommands = new Command[3];
    private Command[] offCommands = new Command[3];
    private Stack<Command> history = new Stack<>();

    public void setOnCommand(Command command, int slot){
        onCommands[slot] = command;
    }

    public void setOffCommand(Command command, int slot){
        offCommands[slot] = command;
    }

    public void pressOnButton(int slot) throws CloneNotSupportedException{
        onCommands[slot].execute();
        history.push(onCommands[slot].clone());
    }

    public void pressOffButton(int slot) throws CloneNotSupportedException{
        offCommands[slot].execute();
        history.push(offCommands[slot].clone());
    }

    public void pressUndo() throws CloneNotSupportedException{
        Command command = null;
        boolean isActed = false;
        while(!history.empty() && !isActed){
            command = history.pop();
            isActed = command.undo();
        }
    }
}

File: CommandPattern/python/RemoteControl.py
import copy
from Command import *

class RemoteControl:
    def __init__(self) -> None:
        self._onCommands:list[Command] = [None, None, None]
        self._offCommands:list[Command] = [None, None, None]
        self._history:list[Command] = list()

    def setOnCommand(self, command:Command, slot:int) -> None:
        self._onCommands[slot] = command

    def setOffCommand(self, command:Command, slot:int) -> None:
        self._offCommands[slot] = command

    def pressOnButton(self, slot:int) -> None:
        self._onCommands[slot].execute()
        self._history.append(copy.copy(self._onCommands[slot]))

    def pressOffButton(self, slot:int) -> None:
        self._offCommands[slot].execute()
        self._history.append(copy.copy(self._offCommands[slot]))

    def pressUndo(self) -> None:
        isActed = False
        while self._history and not isActed:
            command = self._history.pop()
            isActed = command.undo()


